### ISSUE TITLE
Add Select Accessibility UX design discussion

### ISSUE THREAD

**Issue by davidmenendez** [carbon-design-system/ibm-products#2715]
After a recent conversation with our developer @janhassel, it was discovered that there were several major accessibility issues related to the add select pattern. When asked for his notes during the add select accessibility review he had this to say

>Hey, I just took a look. I don’t think the current semantic setup is working for either the SingleAddSelect nor the MultiAddSelect.
>
> The SingleAddSelect currently uses ungrouped radio buttons. A radio button can not have any interactive elements tied to it though (like the “view children” button) and as soon as there is more than more than one item with children or it’s not the last it doesn’t work from a keyboard navigation standpoint either.
>
> The MultiAddSelect currently uses ungrouped checkboxes. There is no semantic relation between the checkboxes and their actions (like “view children”), their children or their adjacent checkboxes on the same level. In the “with modifiers” story I can also not identify any relation to the associated controls (like the role dropdown).
>
> I think the pattern that make the most sense here would be a treegrid. It brings in all features of a data grid role (single- and multi-selection, interactive elements within a row, sorting, etc.) as well as allowing for hierarchical structures (like a treeview).
>
> Here’s the APG write-up of a treegird: https://www.w3.org/WAI/ARIA/apg/patterns/treegrid/
> 
> I believe the multiple columns on the “MultiSelect/With Hierarchy” can be achieved in the treegrid with aria-owns.
>
> I understand that this may require some ground work but I honestly don’t see any way to make it work with the current semantics and I don’t think it makes sense to start an a11y review based on the current structure as 95% of all points would be irrelevant for a treegrid setup.
>
> I would have some time tomorrow or on Monday (during Katrina’s 1Q design review I guess) if you wanna chat about it! :slightly_smiling_face:

In a follow up asking about wether or not he believed these changes could be done incrementally and still allow add select to be released he had this to say

> I totally understand that and agree that a thorough accessibility review / check-in session should’ve happened a lot earlier. I’m not the one making the call ultimately but if I would be – yes, I see this as ship blocker for the following reason:
>
>As you say yourself, this could/will be a lot of work and likely requires a good chunk of DOM changes which can easily trigger a breaking change which in turn would mean that it can’t be fixed until the next major release of the library. And in the current state it does not meet accessibility requirements and can be a literal deal breaker for customers/business partners.

So, as it stands, it looks like a serious look needs to be taken at add select both from a design and dev perspective. The main problem seems to be related to the hierarchy view, which adds the view children button within the radial button, as well as any other functionality that adds suplimental buttons such as view meta data.

I don't know if this will require an entirely new design or teardown of code, but we definitely need to take a look at this to ensure accessibility standards are met. I'd like to open it up to comments from dev, design, and accessibility. If I have not tagged anyone who you believe is pertinent to this conversation please tag them and if we need to communicate on slack or meet on webex we can certainly plan to do so.

@janhassel 
@elycheea 
@kristastarr

**Comment by davidmenendez** [comment]
@OliverKauselmann
@marion-bruells

**Comment by elycheea** [comment]
Let’s take a look at this from the`treegrid` perspective as Jan suggested. We can see how close we can get it to the original design using the role and followup with design once we have basic prototype of it.

**Comment by davidmenendez** [comment]
i have received a message from Oliver on slack and he and Jan are looking at the issue. luckily they're both co located in the germany office, so hopefully they can work together quickly to find a solution.

**Comment by davidmenendez** [comment]
i'm going to close this issue since the major offenders were taken care of, but i will open another issue to track some of the additional smaller issues that also need to be addressed.
